### PR TITLE
Dashboard: Migrating dashboard settings to react

### PR DESCRIPTION
**Pull request by torkelo**
Lots remaining but wanted to get this going 

Todo 
* [x] Sidenav and moving between subpages
* [x] General settings
* [x] Annotations editor
* [x] Variables editor  (was already in react)
* [x] Loading permissions react view works
* [x] Update DashboardPermissions.tsx to return named export (not default, see DashboardPanel.tsx for example), replace connectWithStore with normal connect
* [x] Links
* [x] History
* [x] JSON model 
* [x] JSON model Save changes only when dashboard.meta.canEdit is true 
* [x] Make editable flow 
* [x] Prevent showing dashboard permissions view after unsaved folder change 
* [x] Save / SaveAs button 

load dash links editor: 
```typescript
componentDidMount() {
    const loader = getAngularLoader();

    const template = '<dash-links-editor dashboard="ctrl.dashboard"></dash-links-editor>';
    const scopeProps = { dashboard: this.props.dashboard };
    this.angularCmp = loader.load(this.element, scopeProps, template);
  }
```

load history directive: 
```typescript
componentDidMount() {
    const loader = getAngularLoader();

    const template = '<gf-dashboard-history dashboard="dashboard"></gf-dashboard-history>';
    const scopeProps = { dashboard: this.props.dashboard };
    this.angularCmp = loader.load(this.element, scopeProps, template);
  }
```

### CODE REVIEW

**Comment by torkelo** [comment]
@jackw 

Tested this and could not find any bug! Great job! Can't approve it as it's my PR, but from my POV it's approved :)

File: public/app/core/angular_wrappers.ts [hunk 1]
```diff
@@ -200,7 +199,6 @@ export function registerAngularDirectives() {
     ['getDashboard', { watchDepth: 'reference', wrapApply: true }],
     ['onSaveSuccess', { watchDepth: 'reference', wrapApply: true }],
   ]);
-  react2AngularDirective('variableEditorContainer', VariableEditorContainer, []);
```
Comment (hugohaggmark) [inline public/app/core/angular_wrappers.ts:203]: Horray! 🎉

File: public/app/features/dashboard/components/DashboardPermissions/DashboardPermissions.tsx [hunk 2]
```diff
@@ -1,34 +1,42 @@
...
+
+export interface DispatchProps {
   getDashboardPermissions: typeof getDashboardPermissions;
   updateDashboardPermission: typeof updateDashboardPermission;
   removeDashboardPermission: typeof removeDashboardPermission;
   addDashboardPermission: typeof addDashboardPermission;
 }
 
+export type Props = OwnProps & ConnectedProps & DispatchProps;
+
```
Comment (hugohaggmark) [inline public/app/features/dashboard/components/DashboardPermissions/DashboardPermissions.tsx:34]: As you added Own & ConnectedProps maybe you should follow the guide here too: https://github.com/grafana/grafana/blob/master/contribute/style-guides/redux.md#typing-of-connected-props

File: public/app/types/folders.ts [hunk 3]
```diff
@@ -23,7 +23,7 @@ export interface FolderState {
 }
 
 export interface FolderInfo {
-  id: number;
-  title: string;
-  url: string;
+  id: number | undefined;
+  title: string | undefined;
+  url: string | undefined;
```
Comment (hugohaggmark) [inline public/app/types/folders.ts:28]: nit: why not use `?`
```suggestion
  id?: number;
  title?: string;
  url?: string;
```

**Review by hugohaggmark — APPROVED** [review]
Substantial work on this Torkel & Jack ⭐  I tried it out, and it works as expected found this small nit in the Versions settings:
image

File: public/app/features/dashboard/components/DashboardPermissions/DashboardPermissions.tsx (hunk 2 above)
Comment (jackw) [reply public/app/features/dashboard/components/DashboardPermissions/DashboardPermissions.tsx]: Thanks for the link @hugohaggmark. I've updated this so types are automatically inferred! ⚡

Comment (hugohaggmark) [reply public/app/features/dashboard/components/DashboardPermissions/DashboardPermissions.tsx]: Awesome!

**Pull request merged**